### PR TITLE
[core] add getJSBundleFile support for bridgeless mode

### PR DESCRIPTION
**Pull request by Kudo**
# Why

expo-updates uses the `ReactNativeHost.getJSBundleFile()` to define the launch bundle path. the defaultReactHost doesn't support it, so expo-updates always uses the embedded bundle.

# How

reference the logic from bridged ReactInstanceManager to support the `getJSBundleFile()`: https://github.com/facebook/react-native/blob/15a5638c621cbec4a9fcb5ae94938120cdc32fae/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/ReactInstanceManagerBuilder.java#L107-L114

# Test Plan

test canary + expo-updates + eas updates that should launch the remote bundle

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [x] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

**Review by lukmccall — APPROVED** [review]

**Review by tsapeta — APPROVED** [review]

File: packages/expo/android/src/main/java/expo/modules/ExpoReactHostFactory.kt [hunk 1]
```diff
@@ -41,6 +41,12 @@ object ExpoReactHostFactory {
     override val jsBundleLoader: JSBundleLoader
       get() {
         val context = weakContext.get() ?: throw IllegalStateException("Unable to get concrete Context")
+        reactNativeHostWrapper.jsBundleFile?.let { jsBundleFile ->
```
Comment (ide) [inline packages/expo/android/src/main/java/expo/modules/ExpoReactHostFactory.kt:44]: Would it be helpful to add comments saying which parts of the code run in bridgeless mode vs. with the bridge enabled? Later, when there is only bridgeless mode, we can remove the comments.

**Review by ide — COMMENTED** [review]
@wschurman adding you as a post-commit reviewer so you're aware of this change. It should be relatively quick to look at.

File: packages/expo/android/src/main/java/expo/modules/ExpoReactHostFactory.kt (hunk 1 above)
Comment (Kudo) [reply packages/expo/android/src/main/java/expo/modules/ExpoReactHostFactory.kt:44]: good question! basically this whole ExpoReactHostFactory file is for bridgeless. react-native has a convention
 - bridgeless: `ReactHost`, `ReactInstance`
 - bridge: `ReactNativeHost`, `ReactInstanceMananger`, `Catalyst*`

there's some effort to adapt from ReactNativeHost to ReactHost without much change. in this example, it [reads some necessary properties from ReactNativeHost](https://github.com/facebook/react-native/blob/ad9f0b0ab88218beb214859e0d85d50f9caa7bb5/packages/react-native/template/android/app/src/main/java/com/helloworld/MainApplication.kt#L32-L33).

basically i would try to follow the same effort to migrate our [ReactNativeHostHandler](https://github.com/expo/expo/blob/main/packages/expo-modules-core/android/src/main/java/expo/modules/core/interfaces/ReactNativeHostHandler.java) to be bridgeless/bridge independent, so that expo modules don't have to handle these different cases. though expo-dev-client and expo-updates are two special cases that they are some tight coupled code inside. we are aware and working on the migration.

Comment (ide) [reply packages/expo/android/src/main/java/expo/modules/ExpoReactHostFactory.kt:44]: Ah okay. Thanks for the explanation. If I understand correctly, does that mean that when bridged mode is removed (one day, in the future), we'll move the code from ReactNativeHostWrapper into ReactHostFactory, since ReactNativeHost will no longer exist?

Comment (Kudo) [reply packages/expo/android/src/main/java/expo/modules/ExpoReactHostFactory.kt:44]: > If I understand correctly, does that mean that when bridged mode is removed (one day, in the future), we'll move the code from ReactNativeHostWrapper into ReactHostFactory, since ReactNativeHost will no longer exist?

right! we will merge into one class. the class name might be ReactHostWrapper, depending on how MainApplication from the template changes over time. if the [logic for ReactNativeHost](https://github.com/expo/expo/blob/3d13e80d72893f7429a237a47ed74be4c5388eb6/templates/expo-template-bare-minimum/android/app/src/main/java/com/helloworld/MainApplication.kt#L20-L36) moved to ReactHost, then we may have a ReactHostWrapper after all.

**Review by wschurman — APPROVED** [review]
